### ISSUE TITLE
0.7.0 Migration Guide

### ISSUE THREAD

**Issue by minghuaw** [minghuaw/fe2o3-amqp#120]
Below shows that 0.7.0 migration guide.

**Comment by minghuaw** [comment]
# 0.7.0 Migration Guide

## TL;DR

### Usages that should remain compatible

1. `Message::builder` should work just fine
2. Sending primitive types (types defined in the first part of the core spec; or the types covered by `serde_amqp::Value`) 

### Usages that might be broken

1. Sending a custom type that implements `Serialize`. Custom type should be wrapped explicitly in a body section type or implement `IntoBody` so it can be converted into a `Sendable`

```rust
#[derive(Serialize)]
struct Foo {
    a: i32
}

sender.send(AmqpValue(Foo{a: 3})).await.unwrap();
```

OR

```rust
#[derive(Serialize)]
struct Foo {
    a: i32
}

impl IntoBody for Foo {
    type Body = AmqpValue<Foo>;

    fn into_body(self) -> Self::Body {
        AmqpValue(self)
    }
}

sender.send(Foo{a: 5}).await.unwrap();
```

2. Receiving `Delivery<Value>` (ie. `receiver.recv::<Value>().await`)

Now that the message body is a generic, the user needs to specify the body section type (one of the types that implement `DeserializableBody`, see below). This makes it more ergonomic if the user is certain about the exact message body type, but slightly more complicated if the body section type is either not known or may change from time to time.

If the exact message body section type is known, the user no longer needs to manually match different message body variants or use the `.try_into_x()` or `.try_as_x()` methods.

```rust 
// Assume the body section will always be an uninterpreted bag of bytes (ie. `Data`)
let delivery = receiver.recv::<Data>().await.unwrap();
// User no longers needs to use `.try_as_data()` method
let data: &Data = delivery.body();
```

If the user is not sure about the exact body section or the type of encoded value (ie. previously `Delivery<Value>`), the user should choose `Body<Value>` as the generic parameter. Please note that `Body::Data` and `Body::Sequence` now contains a vec to allow one or more `Data` and one or more `AmqpSequence` sections.

```rust
let delivery: Delivery<Body<Value>> = receiver.recv::<Body<Value>>().await.unwrap();
match delivery.body() {
    Body::Value(_) => todo!(),
    Body::Sequence(vec) => todo!(),
    Body::Data(vec) => todo!()
}
```

## What has changed?

The biggest changes in this breaking release all revolves around message body (ie. field `body` in the `Message` type), which now looks like

```rust
pub struct Message<B> {
    // other fields
    // ...
    body: B,
    // ...
}
```

The struct doesn't define any trait bound for the generic type `B`. However, `Serialize` or `Deserialize` are only implemented when `B` meets the newly added trait bound `SerializableBody` or `DeserializableBody`. These two traits are simply marker traits that doesn't have any associated methods, and they only act as a constraint that ensures the message body is serialized/deserialized correctly. The following types 

WIP

**Comment by minghuaw** [comment]
### `IntoBody`, `FromBody` and `FromEmptyBody`

As the name suggests, these traits define how a type should be converted into and from message body section(s). 

- `IntoBody` - how to convert a type into a message body section
- `FromBody` - how to convert a type from a message body section
- `FromEmptyBody` - though the spec doesn't allow an empty body, it was later verified by the protocol author that this was a mistake in the wording, so a valid message may have zero body section. This trait defines how an empty body should be interpreted.

They have been implemented for all the primitive types already. If there is a custom struct `T`, one could implement these three traits and then use `T` as the generic parameter for sending and receiving.

```rust
#[derive(Serialize)]
struct Foo {
    a: i32
}

impl IntoBody for Foo {
    type Body = AmqpValue<Foo>;

    fn into_body(self) -> Self::Body {
        AmqpValue(self)
    }
}

impl FromBody for Foo {
    type Body = AmqpValue<Foo>;

    fn from_body(body: Self::Body) -> Self {
        body.0
    }
}

impl FromEmptyBody for Foo {} // Use the blanket implementation

sender.send(Foo{a: 5}).await.unwrap();
```

**Comment by minghuaw** [comment]
### `Batch<Data>` and `Batch<AmqpSequence<T>>`

### `Body::Data(Batch<Data>)` and `Body::Sequence(Batch<AmqpSequence<T>>)`
